### PR TITLE
gotify: 2.0.21 -> 2.1.0

### PR DESCRIPTION
**Pull request by Ma27**
###### Motivation for this change
ChangeLog: https://github.com/gotify/server/releases/tag/v2.0.22
ChangeLog: https://github.com/gotify/server/releases/tag/v2.0.23
ChangeLog: https://github.com/gotify/server/releases/tag/v2.1.0

While the update only contains a few small features and a few bugfixes,
the change was rather messy for us unfortunately:

* It seems as if `npmjs.org`-packages can't be transformed into
  `pkg___pkg-x.y.z` for Yarn's offline cache. The name
  `https___registry.npmjs.org_caniuse_lite___caniuse_lite_1.0.30001237.tgz`
  isn't the problem because when changing the URL "parser" of `yarn2nix`
  to transform this into `org_caniuse_lite___caniuse_lite_1.0.30001237`
  this doesn't help either.

  Instead, I derived the fix from `gitlab`[1] where `yarn.lock` gets
  patched to make sure that it detects the package in the offline-cache
  properly.

* The frontend is now built with `react-scripts`. This is a problem for
  us because it tries to write into `node_modules/.cache` even though
  `node_modules` is a store-path in the context of `yarn2nix`[2].

  The change isn't pretty, but solves the issue for us.

[1] https://github.com/NixOS/nixpkgs/blob/f007b794c758000a275b00dd0695d2fb155195f0/pkgs/applications/version-management/gitlab/default.nix#L85-L86
[2] https://github.com/facebook/create-react-app/issues/11263

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/#sec-conf-file))
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/servers/gotify/ui.nix [hunk 1]
```diff
@@ -10,16 +10,52 @@ yarn2nix-moretea.mkYarnPackage rec {
...
+      owner = "gotify";
+      repo = "server";
+      rev = "v${version}";
+      sha256 = import ./source-sha.nix;
+    };
+    postPatch = ''
+      substituteInPlace ui/yarn.lock \
+        --replace \
+          "https://registry.npmjs.org/caniuse-lite/-/caniuse-lite-1.0.30001237.tgz" \
+          "https___registry.npmjs.org_caniuse_lite___caniuse_lite_1.0.30001237.tgz"
```
Comment (doronbehar) [inline pkgs/servers/gotify/ui.nix:24]: I'm not sure I understood what's the cause for this issue, but perhaps it should be reported to `yarn2nix` project or to `yarn`?

File: pkgs/servers/gotify/ui.nix [hunk 2]
```diff
@@ -10,16 +10,52 @@ yarn2nix-moretea.mkYarnPackage rec {
...
   src = "${src_all}/ui";
 
   buildPhase = ''
+    export HOME=$(mktemp -d)
+    export WRITABLE_NODE_MODULES="$(pwd)/tmp"
+    mkdir -p "$WRITABLE_NODE_MODULES"
+
+    # react-scripts requires a writable node_modules/.cache, so we have to copy the symlink's contents back
+    # into `node_modules/`.
+    # See https://github.com/facebook/create-react-app/issues/11263
```
Comment (doronbehar) [inline pkgs/servers/gotify/ui.nix:36]: This comment is good but perhaps should be at the beginning of the `buildPhase`.

**Review by doronbehar — COMMENTED** [review]
Thanks for maintaining this package @Ma27 !

File: pkgs/servers/gotify/ui.nix (hunk 1 above)
Comment (Ma27) [reply pkgs/servers/gotify/ui.nix:24]: I think that yarn wrongly recognizes it as some kind of external dependency, so we override it with the `___`-style that's used for the artificial offline cache.

However I don't think that yarn really cares about people trying to re-implement this somehow.

File: pkgs/servers/gotify/ui.nix (hunk 2 above)
Comment (Ma27) [reply pkgs/servers/gotify/ui.nix:36]: There are a few more comments related to the blocks separated by whitespace, so I think it's OK to to leave it as-is, right? :)

**Pull request merged**